ruby_library/entities/library.rb

```diff
@@ -0,0 +1,100 @@
+require_relative '../storage/storage'
+
+class Library
+    include Storage
+
+    def initialize(authors=[], books=[], readers=[], orders=[])
+        @authors = []
+        @books = []
+        @readers = []
+        @orders = []
+        add(authors, books, readers, orders)
+    end
+
+    def to_s
+        "Library\n" + 
+        "Authors:\n\t#{@authors.join("\n\t")}\n" + 
+        "Books:\n\t#{@books.join("\n\t")}\n" +
+        "Readers:\n\t#{@readers.join("\n\t")}\n" + 
+        "Orders:\n\t#{@orders.join("\n\t")}\n"
+    end
+
+    def to_h
+        {
+         authors: @authors,
+         books: @books,
+         readers: @readers,
+         orders: @orders
+        }
+    end
+
+    def add(*entities)
+        entities.flatten!
+        entities.each do |entity|
+            case entity
+            when Author
+                @authors.push(entity) unless @authors.include?(entity)
+            when Book
+                @books.push(entity) unless @books.include?(entity)
+                @authors.push(entity.author) unless @authors.include?(entity.author)
+            when Reader
+                @readers.push(entity) unless @readers.include?(entity)
+            when Order
+                if !@orders.include?(entity) && @readers.include?(entity.reader) && @books.include?(entity.book)
+                    @orders.push(entity)
+                end
+            end
+        end
+    end
+
+    alias + add
+
+    def save(file)
+        save_yml(to_h, file)
+    end
+
+    def load(file)
+        h = load_yml(file)
+        if h.is_a?(Hash) && h.keys == [:autors, :books, :readers, :orders]
+            @authors = h[:authors]
+            @books = h[:books]
+            @readers = h[:readers]
+            @orders = h[:orders]
+        end
+    end
+
+    def top_book
+        top_books(1).first
+    end
+
+    def top_books(n=3)
+        top_entities(n, @books, :@book)
+    end
+
+    def top_readers(n=3)
+        top_entities(n, @readers, :@reader)
+    end
+
+    def count_of_top_books_readers(n=3)
+        books = top_books(n)
+        readers = []
+        @orders.each do |order|
+            if books.include?(order.book) && !readers.include?(order.reader)
+                readers.push(order.reader)
+            end
+        end
+        readers.size
+    end
+
+    private
+
+    def top_entities(n=3, entities, variable)
```

Comment:
try to rethink the logic of that method
1) named parameters use
2) n -> quantity
3) Method is too complex. Use `group_by` for grouping hash by some parameter. This method must contain at most - 4 lines

Reply:
ok